File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        Arrays.sort(jobs, (a, b) -> Integer.compare(a[0], b[0]));
        int answer = 0;
        int n = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            if (a[0] != b[0]) return Integer.compare(a[0], b[0]);   
            if (a[1] != b[1]) return Integer.compare(a[1], b[1]);  
            return Integer.compare(a[2], b[2]);                    
        });
        int time = 0;
        
        while (n < jobs.length || pq.size() != 0) {
            while (true) {
                if (jobs.length > n && time >= jobs[n][0]) {
                    pq.add(new int[]{jobs[n][1], jobs[n][0], n});
                    n += 1;
                } else break;
            }
            
            if (pq.size() == 0) {
                time += 1;
                continue;
            }
            
            int[] item = pq.poll();
            time += (item[0]);
            answer += (time - item[1]);
        }
        return answer / jobs.length;
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
from queue import PriorityQueue

def solution(jobs):
    jobs.sort(key=lambda x: x[0])
    answer = 0
    n = 0
    queue = PriorityQueue()
    time = 0
    
              
    while len(jobs) != 0 or not queue.empty():
        while True:
            if jobs and time >= jobs[0][0]:
                item = jobs.pop(0)
                queue.put((item[1], item[0], n))
                n += 1
            else:
                break
                
        if queue.empty():
            time += 1
            continue
        
        item = queue.get()
        time += item[0]
        answer += (time - item[1])
        
    return answer // n